IBX-1430: Revise REST API documentation for 4.x (#1591)
ezsystems/developer-documentation: docs/api/rest_api_authentication.md

This page refers to [REST API reference](rest_api_reference/rest_api_reference.html), where you can find detailed information about
REST API resources and endpoints.

Five authentication methods are currently supported: session (default), JWT, basic, OAuth and client certificate (SSL).

> Those methods can't be used at the same time.

Author: This is true. I started a priority contest; which wins when enable at the same time. I establish "JWT > Basic > Session" but did go further as it was useless. The only relevant information is that only one work so only one should be enabled.

- [Cross-origin requests](rest_api_usage.md#cross-origin-requests)
- [`access_control`]([[= symfony_doc =]]/security/access_control.html)

## Session-based authentication

This authentication method requires a Session cookie to be sent with each request.

If this authentication method is used with a web browser, this session cookie is automatically available as soon as your visitor logs in.
Add it as a cookie to your REST requests, and the user will be authenticated.

Sessions are created to re-authenticate the user only (and perform authorization), not to hold session state in the service.
Because of that, we regard this method as supporting AJAX-based applications even if it violates the principles of RESTful services.

### Configuration

Session is the default method and is already enabled; no configuration required. Notice that enabling whatever other method will disable session.

### Usage examples

You can create a session for a visitor even if they are not logged in by sending the **`POST`** request to `/user/sessions`.
For logging out, use the **`DELETE`** request on the same resource.

#### Establishing a session

##### Creating session

> 

Author: Not really inspired. Some like "First, the session itself must be created before, second, being fully authenticated into it."?

> 

Reviewer: Even a simpler intro like: "To create a session, execute the following REST request:"

> 

Author: ```suggestion

To create a session, execute the following REST request:

```

      <User href="/user/users/14" media-type="vnd.ibexa.api.User+xml"/>
    </Session>
    ```

=== "JSON"

    ```
    POST /user/sessions HTTP/1.1
    Host: www.example.net
    Accept: application/vnd.ibexa.api.Session+json
    Content-Type: application/vnd.ibexa.api.SessionInput+json
    Cookie: eZSESSID98defd6ee70dfb1dea416=go327ij2cirpo59pb6rrv2a4el2
    X-CSRF-Token: 23lk.neri34ijajedfw39orj-3j93
    ```
    
    ```xml
    {
      "SessionInput": {
        "login": "admin",
        "password": "publish"
      }
    }
    ```
    
    ```
    HTTP/1.1 200 OK
    Content-Type: application/vnd.ibexa.api.Session+json
    ```
    
    ```xml
    {
      "Session": {
        "_media-type": "application\/vnd.ibexa.api.Session+json",
        "_href": "\/api\/ibexa\/v2\/user\/sessions\/jg1nhinvepsb9ivd10hbjbdp4l",
        "name": "eZSESSID98defd6ee70dfb1dea416",
        "identifier": "go327ij2cirpo59pb6rrv2a4el2",
        "csrfToken": "23lk.neri34ijajedfw39orj-3j93",
        "User": {
          "_media-type": "application\/vnd.ibexa.api.User+json",
          "_href": "\/api\/ibexa\/v2\/user\/users\/14"
        }
      }
    }
    ```

#### Using the session

##### Session cookie

> The cookie previously set can now be joined to request to be executed with this user.

Author: I'll remove the second "previously"
```suggestion
The cookie previously set can now be joined to request to be executed with this user.
```

### Configuration

To enable HTTP basic authentication, edit `config/packages/security.yaml`, and, before the `ibexa_front` firewall, add a new firewall dedicated to REST with [`http_basic`](https://symfony.com/doc/5.4/security.html#http-basic) configuration:

```yaml
        ibexa_rest:
            pattern: ^/api/ibexa/v2
            http_basic:

>                 realm: Ibexa DXP REST API

Author: Fixed with a dedicated host circa b9116e19a8e481603da1da620ca0ba302379024f

    
    ```
    GET /content/locations/1/5/children
    Host: <yourdomain>
    Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9…-QBE4-6eKNjg
    Accept: application/vnd.ibexa.api.LocationList+json
    ```

## HTTP basic authentication

For more information, see [HTTP Authentication: Basic and Digest Access Authentication](http://tools.ietf.org/html/rfc2617).

### Configuration

If the installation has a dedicated host for REST, you can enable HTTP basic authentication only on this host by setting a firewall like the following before the `ibexa_front` one:

```yaml
        ibexa_rest:
            host: ^api\.example\.com$
            http_basic:
                realm: Ibexa DXP REST API
```

!!! caution "Back Office uses REST API"

    Notice that the Back Office uses the REST API too (for some parts like the Location tree or the Calendar) on its own domain.
    
    * If the Back Office SiteAccess matches admin.example.com, it will call the REST API under //admin.example.com/api/ibexa/v2;
    * If the Back Office SiteAccess matches localhost/admin, it will call the REST API under //localhost/api/ibexa/v2.
    

>     If Basic authentication is aimed only for REST API, it is better to have a dedicated domain even on a developpement environement.

Reviewer: ```suggestion
    If basic authentication is used only for REST API, it is better to have a dedicated domain even on a development environment.
```

### Usage example

Basic authentication requires the username and password to be sent *(username:password)*, based 64 encoded, with each request.
For details, see [RFC 2617](http://tools.ietf.org/html/rfc2617).

> Most HTTP client libraries as well as REST libraries support this method. [Creating content with binary attachments](rest_api_usage.md#creating-content-with-binary-attachments) has an example using Basic authentication with [cURL](https://www.php.net/manual/en/book.curl.php) and its `CURLOPT_USERPWD`. 

Reviewer: ```suggestion
Most HTTP client libraries as well as REST libraries support this method.
[Creating content with binary attachments](rest_api_usage.md#creating-content-with-binary-attachments) has an example using basic authentication with [cURL](https://www.php.net/manual/en/book.curl.php) and its `CURLOPT_USERPWD`. 
```
